### ISSUE TITLE
Stop _AttemptManager from using Queue.noisy

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6705]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)| @staceysern reported|
|-|-|
|Trac ID|trac#6705|
|Type|enhancement|
|Created|2013-08-27 14:48:05Z|
|Branch|https://github.com/twisted/twisted/tree/attemptmanager-noisy-6705-2|

Ticket [#4623](https://github.com/twisted/twisted/issues/4623) points out that relaymanager._AttemptManager directly uses relaymanager.Queue.noisy.  So as not to include .noisy in the new IRelayQueue interface, _AttemptManager should not use that attribute.

<details><summary>Searchable metadata</summary>

```
trac-id__6705 6705
type__enhancement enhancement
reporter__shira shira
priority__normal normal
milestone__ 
branch__branches_attemptmanager_noisy_6705_2 branches/attemptmanager-noisy-6705-2
branch_author__shira shira
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1377614885000000 1377614885000000
changetime__1421493084169395 1421493084169395
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_1"></a>|@staceysern commented|
|-|-|

(In [39782]) Branching to 'attemptmanager-noisy-6705'

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_2"></a>|@staceysern commented|
|-|-|

(In [39981]) Initial commit

refs [#6705](https://github.com/twisted/twisted/issues/6705)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_3"></a>|@staceysern commented|
|-|-|

(In [39982]) Fix twistedchecker errors

refs [#6705](https://github.com/twisted/twisted/issues/6705)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)|@staceysern removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl set owner to @staceysern|
|-|-|

Hi shira, thanks for working on this!

1. (Line 2229, test_mail.py) "realyed" should be "relayed"

2. The epytext for Queue should be defined in its __init__ function, not in the class. (Line 2242, test_mail.py)

3. Fullstops required on line 2261, 2243 in test_mail.py.

Minor things which I think might be worth fixing while you're touching code nearby, but is your call:

1. There is incorrect indentation on line 325.

2. Lines 314 and 315 could probably be pulled together. I'm also not sure that it actually does as it says? Maybe I'm just not understanding the code.

3. Line 320 is too long.

4. _AttemptManager has incorrect spacing above it.

Please fix/address these, and the minor things if you think the scope of this ticket is appropriate, and resubmit. FWIW, I can't seem to find anything else wrong, other than those two small typo points, and the builds are all green.

PS: Unrelated to the review, line 2363 of test_mail.py (`self.assertTrue(am.noisy)`) did make me giggle :)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_7"></a>|@staceysern commented|
|-|-|

(In [40051]) Incorporate review comments

refs [#6705](https://github.com/twisted/twisted/issues/6705)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_8"></a>|@staceysern removed owner|
|-|-|

I fixed major items 1-3.  I also fixed minor items 1 and 3.  Item 4 will be fixed as part of ticket [#6739](https://github.com/twisted/twisted/issues/6739).  

Regarding minor item #2, I agree that what the comment says is not implemented.  Ticket [[#6726](https://github.com/twisted/twisted/issues/6726)](https://github.com/twisted/twisted/issues/6726) documents a problem related to that.  I did not change the comment since it should be addressed by #6726.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

Hi shira, thanks! I'll leave someone else to review it now, just in case I've overlooked anything.

RE: [#6739](https://github.com/twisted/twisted/issues/6739), feel free to assign it to me when you put it up for review, and I'll take a look.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_11"></a>|@lvh set owner to @staceysern|
|-|-|

Hi shira,


Thanks for working on this some more :) This is one of those great tickets where the state of the code after the merge is so much better than the state before :)

Small notes (not really review points):

 - If [the reactor isn't passed](https://github.com/twisted/twisted/compare/trunk...attemptmanager-noisy-6705#diff-c8c8e73bb6382f73aba4b74e11a9f2dbR253), the code imports the global reactor as globalReactor, and then assigns reactor to it. globalReactor doesn't appear to be used anywhere else. Is that intentional? From what I can tell just ```from twisted.internet import reactor``` would do the same thing.
 - We may at some point want to use [#6365](https://github.com/twisted/twisted/issues/6365) for this instead of yet another manual deferred notification thing, but it seems that won't be closed before this is.
 - I'm unfamiliar with mail, but in the docstring for ```def notifySuccess(self, relay, message):``` it may be good to clarify that you're removing it from the queue (that clarified it for me at least)
 - I'm not a huge fan of the ```from_``` and ```to``` variable names (how about sender, receiver?) but I think they're used all over t.mail already so maybe the damage done from inconsistency is greater than any naming improvement :)
 - Is the hardcoded 30 second wait time in ```notifyNoConnection``` intentional?
 - The "Set up" wording of the ```notify*``` calls is kind of confusing to me, I had to re-read the docstring a few times before I grokked it. Perhaps "prepare"?
 - There doesn't appear to be a difference between "done" as in ```notifyDone``` and "completed" as in ```getCompletionDeferred```. Is that intentional?

I did find a bug in one of the docstrings:

 - ```notifyDone``` claims to fire a Deferred, but it actually fires *all* of the notification deferreds. Perhaps a reference to ```getCompletionDeferred```?

Other than that, looks great to me as well as the buildbots. I don't know if any of my nitpicks hold any water, I'm more than happy to merge this as-is modulo the minor docstring bug :)

cheers

lvh

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_12"></a>|@staceysern commented|
|-|-|

(In [43028]) Branching to 'attemptmanager-noisy-6705-2'

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_13"></a>|@staceysern commented|
|-|-|

(In [43030]) Remove conflict
refs [#6705](https://github.com/twisted/twisted/issues/6705)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_14"></a>|@staceysern commented|
|-|-|

(In [43031]) Address review comments

refs [#6705](https://github.com/twisted/twisted/issues/6705)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_15"></a>|@staceysern commented|
|-|-|

(In [43033]) Remove trailing whitespace

refs [#6705](https://github.com/twisted/twisted/issues/6705)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_16"></a>|@staceysern commented|
|-|-|

(In [43034]) Improve docstring

refs [#6705](https://github.com/twisted/twisted/issues/6705)

**Comment by twisted-trac** [comment]
|[<img alt="staceysern's avatar" src="https://avatars.githubusercontent.com/u/3631162?s=50" width="50" height="50">](https://github.com/staceysern)<a name="note_17"></a>|@staceysern set owner to @glyph|
|-|-|

I addressed most of the issues in comment:11.  I do not know if either the hard-coded 30 seconds or the difference between done and complete are intentional.  Regarding the 30 seconds, I can't figure out but would like to understand why the calls to setWaiting are delayed.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph removed owner|
|-|-|

Un-assigning all review tickets assigned to me.  Unless there's a specific reason _I_ need to review something, let's leave them unassigned :-).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl set owner to @hawkowl|
|-|-|

This looks good to me except for a minor whitespace thing which I'll fix.

Will merge. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_20"></a>|@hawkowl set status to `closed`|
|-|-|

(In [43683]) Merge attemptmanager-noisy-6705-2: Stop _AttemptManager from using Queue.noisy

Author: shira
Reviewers: hawkowl, lvh
Fixes: [#6705](https://github.com/twisted/twisted/issues/6705)
